### PR TITLE
Feature/hooks

### PR DESCRIPTION
**Pull request by Remi749**
Support to add Hooks to the provisioning. Both GET/POST HTTP requests should be supported.

This will enable customers to start GET/POST HTTP requests at the end of the project setup process.
- Ex: Trigger API's, Flows, Logic Apps, Azure Functions, etc...
- Support pending requests (202)

Example:
```json
{
  "Hooks": [
    {
      "Title": "Get something from API: X",
      "Url": "https://example.com/get/X",
      "Method": "GET",
      "Headers": {
        "Content-Type": "application/json",
        "Cache-Control": "no-cache"
      }
    },
    {
      "Title": "Trigger logic app: X",
      "Url": "https://example.com/trigger/X",
      "Method": "POST",
      "Headers": {
        "Content-Type": "application/json",
        "Cache-Control": "no-cache"
      },
      "Body": { 
        "Body goes here"
      }
    }
  ]
}
```

Relevant issue:
#1

### CODE REVIEW

File: src/webprovisioner.ts [hunk 1]
```diff
@@ -53,10 +54,12 @@ export class WebProvisioner {
     handlers?: string[],
     progressCallback?: (message: string) => void
   ): Promise<any> {
+    console.log('applyTemplate, template:', template, handlers)
+
     Logger.log({
       message: `${this.config.logging.prefix} (WebProvisioner): (applyTemplate): Applying template to web`,
       data: { handlers },
-      level: LogLevel.Info
+      level: LogLevel.Warning
```
Comment (olemp) [inline src/webprovisioner.ts]: Why is _Applying template to web_ warning level?

File: src/webprovisioner.ts [hunk 2]
```diff
@@ -53,10 +54,12 @@ export class WebProvisioner {
     handlers?: string[],
     progressCallback?: (message: string) => void
   ): Promise<any> {
+    console.log('applyTemplate, template:', template, handlers)
```
Comment (olemp) [inline src/webprovisioner.ts]: What's this?

File: src/webprovisioner.ts [hunk 3]
```diff
@@ -72,6 +75,8 @@ export class WebProvisioner {
       }
     )
 
+    console.log('operations', operations)
```
Comment (olemp) [inline src/webprovisioner.ts]: Remove this?

File: src/webprovisioner.ts [hunk 4]
```diff
@@ -103,7 +108,9 @@ export class WebProvisioner {
    * @param config - Provisioning config
    */
   public setup(config: IProvisioningConfig): WebProvisioner {
+    
     this.config = config
+    console.log('setup, Sets up the web provisioner, config:', config)
```
Comment (olemp) [inline src/webprovisioner.ts]: Remove this?

File: src/handlers/exports.ts [hunk 5]
```diff
@@ -40,5 +42,6 @@ export const DefaultHandlerSort: TypedHash<number> = {
   Navigation: 9,
   PropertyBagEntries: 8,
   WebSettings: 10,
-  SiteFields: 0
+  SiteFields: 0,
+  Hook: 11
```
Comment (olemp) [inline src/handlers/exports.ts]: Should be `Hooks` no?

File: .vscode/tasks.json.old [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+    "version": "0.1.0",
+    "command": "gulp",
+    "isShellCommand": true,
```
Comment (olemp) [inline .vscode/tasks.json.old]: We don't need this file I think?

**Review by olemp — CHANGES_REQUESTED** [review]

File: src/webprovisioner.ts (hunk 1 above)
Comment (Remi749) [reply src/webprovisioner.ts]: No idea why I changed this, will fix

File: src/webprovisioner.ts (hunk 4 above)
Comment (Remi749) [reply src/webprovisioner.ts]: All these console.log are old test codes that I forgot about, wierd that linting didn't fix. Will fix all above.

**Comment by Remi749** [comment]
Updated PR

File: src/handlers/hooks.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            throw new Error(`${(result ? ` | ${result} \n\n` : '')}- Hook ${index + 1}/${hooks.length}: ${hook.Title}`)
+          }
+        }))
+      } else if (hook.Method === 'POST') {
+        super.log_info(
+          'processHooks',
+          `Starting POST request: '${hook.Title}'.`
+        )
+
+        hook.Body['pp_webUrl'] = web['_parentUrl']
```
Comment (olemp) [inline src/handlers/hooks.ts:57]: How does this work? The Body in the schema is set to `string` type.

**Review by olemp — CHANGES_REQUESTED** [review]

File: src/handlers/hooks.ts (hunk 7 above)
Comment (Remi749) [reply src/handlers/hooks.ts:57]: Should probably be: { [key: string]: any } instead or maybe use Body interface is that works better. I'll check and update this

Comment (Remi749) [reply src/handlers/hooks.ts:57]: Body worked! updated PR

**Comment by Remi749** [comment]
Updated PR

File: src/schema.ts [hunk 8]
```diff
@@ -131,6 +132,14 @@ export interface IWebSettings {
   [key: string]: string | boolean
 }
 
+export interface IHooks {
+  Title?: string
+  Url: string
+  Method: string
+  Headers?: Headers
+  Body?: Body
```
Comment (olemp) [inline src/schema.ts:140]: This is a default type in TypeScript, right?

Comment (Remi749) [reply src/schema.ts:140]: Yes, same with Headers.

**Pull request merged**